server/src/main/java/org/apache/iotdb/db/query/timegenerator/EngineTimeGenerator.java

```diff
@@ -51,12 +54,17 @@ private void initNode(QueryContext context) throws StorageEngineException {
 
   @Override
   public boolean hasNext() throws IOException {
-    return operatorNode.hasNext();
+    return (cacheTimes != null && cacheTimes.hasMoreData()) || operatorNode.hasNext();
   }
 
   @Override
   public long next() throws IOException {
```

Comment:
how about support a nextTimes method?

Reply:
no, that would make `AndNode` and `OrNode` hard to write